<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scvi-tools CITE-seq tutorial in R, using serialized tutorial components • scviR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scvi-tools CITE-seq tutorial in R, using serialized tutorial components">
<meta property="og:description" content="scviR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scviR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scviR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/citeseq_tut.html">scvi-tools CITE-seq tutorial in R, using serialized tutorial components</a>
    </li>
    <li>
      <a href="../articles/compch12.html">Comparing totalVI and OSCA book CITE-seq analyses</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/scviR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="citeseq_tut_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scvi-tools CITE-seq tutorial in R, using serialized tutorial components</h1>
                        <h4 data-toc-skip class="author">Follows scvi-tools doc, Gayoso, Steier et al. DOI 10.1038/s41592-020-01050-x</h4>
            
            <h4 data-toc-skip class="date">September 26, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/scviR/blob/HEAD/vignettes/citeseq_tut.Rmd" class="external-link"><code>vignettes/citeseq_tut.Rmd</code></a></small>
      <div class="hidden name"><code>citeseq_tut.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="a-cite-seq-example">A CITE-seq example<a class="anchor" aria-label="anchor" href="#a-cite-seq-example"></a>
</h2>
<p>The purpose of this vignette is to illustrate the feasibility of reflecting the material in the <a href="https://colab.research.google.com/github/scverse/scvi-tutorials/blob/0.20.0/totalVI.ipynb" class="external-link">online tutorial</a> for scvi-tools 0.20.0 in Bioconductor. The authors of the tutorial describe it as producing</p>
<blockquote>
<p>a joint latent representation of cells, denoised data for both protein and RNA</p>
</blockquote>
<p>Additional tasks include</p>
<blockquote>
<p>integrat[ing] datasets, and comput[ing] differential expression of RNA and protein</p>
</blockquote>
<p>The integration concerns the simultaneous analysis of two datasets from 10x genomcs.</p>
<p>In this vignette we carry out the bulk of the tutorial activities using R and Bioconductor, reaching to scvi-tools python code via basilisk.</p>
<div class="section level3">
<h3 id="retrieval-of-pbmc-data">Retrieval of PBMC data<a class="anchor" aria-label="anchor" href="#retrieval-of-pbmc-data"></a>
</h3>
<p>The following chunk will acquire (and cache, using BiocFileCache) a preprocessed version of the 10k and 5k combined CITE-seq experiments from the scvi-tools data repository.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/scviR" class="external-link">scviR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="va">adref</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCiteseq5k10kPbmcs.html">getCiteseq5k10kPbmcs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">adref</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 10849 × 4000</span></span>
<span><span class="co">##     obs: 'n_genes', 'percent_mito', 'n_counts', 'batch', '_scvi_labels', '_scvi_batch'</span></span>
<span><span class="co">##     var: 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'</span></span>
<span><span class="co">##     uns: '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'log1p'</span></span>
<span><span class="co">##     obsm: 'protein_expression'</span></span>
<span><span class="co">##     layers: 'counts'</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="retrieval-of-fitted-vae">Retrieval of fitted VAE<a class="anchor" aria-label="anchor" href="#retrieval-of-fitted-vae"></a>
</h3>
<p>The totalVI variational autoencoder was fit with these data. A fitted version is retrieved and cached using</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vae</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCiteseqTutvae.html">getCiteseqTutvae</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB;">INFO    </span> File <span style="color: #BB00BB;">/home/vincent/TEMP/Rtmp8FzFo7/vae2_ov/</span><span style="color: #FF55FF;">model.pt</span> already downloaded </span></span>
<span><span class="co">## <span style="color: #0000BB;">INFO    </span> Computing empirical prior initialization for protein background.</span></span></code></pre>
<p>This is an instance of an S3 class, <code>python.builtin.object</code>, defined in the reticulate package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">vae</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "scvi.model._totalvi.TOTALVI"               </span></span>
<span><span class="co">## [2] "scvi.model.base._rnamixin.RNASeqMixin"     </span></span>
<span><span class="co">## [3] "scvi.model.base._vaemixin.VAEMixin"        </span></span>
<span><span class="co">## [4] "scvi.model.base._archesmixin.ArchesMixin"  </span></span>
<span><span class="co">## [5] "scvi.model.base._base_model.BaseModelClass"</span></span>
<span><span class="co">## [6] "scvi.autotune._types.TunableMixin"         </span></span>
<span><span class="co">## [7] "python.builtin.object"</span></span></code></pre>
<p>Some fields of interest that are directly available from the instance include an indicator of the trained state, the general parameters used to train, and the “anndata” (annotated data) object that includes the input counts and various results of preprocessing:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vae</span><span class="op">$</span><span class="va">is_trained</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="va">vae</span><span class="op">$</span><span class="va">`_model_summary_string`</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## TotalVI Model with the following params: </span></span>
<span><span class="co">## n_latent: 20, gene_dispersion: gene, protein_dispersion: protein, gene_likelihood: nb, latent_distribution: normal</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vae</span><span class="op">$</span><span class="va">adata</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 10849 × 4000</span></span>
<span><span class="co">##     obs: 'n_genes', 'percent_mito', 'n_counts', 'batch', '_scvi_labels', '_scvi_batch'</span></span>
<span><span class="co">##     var: 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'</span></span>
<span><span class="co">##     uns: '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'log1p'</span></span>
<span><span class="co">##     obsm: 'protein_expression'</span></span>
<span><span class="co">##     layers: 'counts'</span></span></code></pre>
<p>The structure of the VAE is reported using</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vae</span><span class="op">$</span><span class="va">module</span></span></code></pre></div>
<p>This is quite voluminous and is provided in an appendix.</p>
</div>
<div class="section level3">
<h3 id="trace-of-negative-elbo-values">Trace of (negative) ELBO values<a class="anchor" aria-label="anchor" href="#trace-of-negative-elbo-values"></a>
</h3>
<p>The negative “evidence lower bound” (ELBO) is a criterion that is minimized in order to produce a fitted autoencoder. The scvi-tools totalVAE elgorithm creates a nonlinear projection of the inputs to a 20-dimensional latent space, and a decoder that transforms object positions in the latent space to positions in the space of observations that are close to the original input positions.</p>
<p>The negative ELBO values are computed for samples from the training data and for “left out” validation samples. Details on the validation assessment would seem to be part of pytorch lightning. More investigation of scvi-tools code and doc are in order.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h</span> <span class="op">=</span> <span class="va">vae</span><span class="op">$</span><span class="va">history</span></span>
<span><span class="va">npts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">h</span><span class="op">$</span><span class="va">elbo_train</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">npts</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">h</span><span class="op">$</span><span class="va">elbo_train</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1200</span>,<span class="fl">1400</span><span class="op">)</span>, </span>
<span>  type<span class="op">=</span><span class="st">"l"</span>, col<span class="op">=</span><span class="st">"blue"</span>, main<span class="op">=</span><span class="st">"Negative ELBO over training epochs"</span>,</span>
<span>  ylab<span class="op">=</span><span class="st">"neg. ELBO"</span>, xlab<span class="op">=</span><span class="st">"epoch"</span><span class="op">)</span></span>
<span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">1360</span>, lty<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"orange"</span><span class="op">)</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"training"</span>, <span class="st">"validation"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">npts</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">h</span><span class="op">$</span><span class="va">elbo_validation</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"l"</span>, col<span class="op">=</span><span class="st">"orange"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citeseq_tut_files/figure-html/lkelb-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="normalized-quantities">Normalized quantities<a class="anchor" aria-label="anchor" href="#normalized-quantities"></a>
</h3>
<p>On a CPU, the following can take a long time.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NE</span> <span class="op">=</span> <span class="va">vae</span><span class="op">$</span><span class="fu">get_normalized_expression</span><span class="op">(</span>n_samples<span class="op">=</span><span class="fl">25L</span>, </span>
<span>    return_mean<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>    transform_batch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBMC10k"</span>, <span class="st">"PBMC5k"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We provide the totalVI-based denoised quantities in</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">denoised</span> <span class="op">=</span> <span class="fu"><a href="../reference/getTotalVINormalized5k10k.html">getTotalVINormalized5k10k</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">denoised</span>, <span class="va">dim</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      rna_nmlzd prot_nmlzd</span></span>
<span><span class="co">## [1,]     10849      10849</span></span>
<span><span class="co">## [2,]      4000         14</span></span></code></pre>
<p>Note that these have features as columns, samples (cells) as rows.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">denoised</span><span class="op">$</span><span class="va">rna_nmlzd</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AL645608.8" "HES4"       "ISG15"      "TTLL10"     "TNFRSF18"  </span></span>
<span><span class="co">## [6] "TNFRSF4"</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">denoised</span><span class="op">$</span><span class="va">prot_nmlzd</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD3_TotalSeqB"  "CD4_TotalSeqB"  "CD8a_TotalSeqB" "CD14_TotalSeqB"</span></span>
<span><span class="co">## [5] "CD15_TotalSeqB" "CD16_TotalSeqB"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="umap-projection-of-leiden-clustering-in-the-totalvi-latent-space">UMAP projection of Leiden clustering in the totalVI latent space<a class="anchor" aria-label="anchor" href="#umap-projection-of-leiden-clustering-in-the-totalvi-latent-space"></a>
</h3>
<p>We have stored a fully loaded anndata instance for retrieval to inspect the latent space and clustering produced by the tutorial notebook procedure. The images produced here do not agree exactly with what I see in the colab pages for 0.20.0. The process was run in Jetstream2, not in colab.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full</span> <span class="op">=</span> <span class="fu"><a href="../reference/getTotalVI5k10kAdata.html">getTotalVI5k10kAdata</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># class distribution</span></span>
<span><span class="va">cllabs</span> <span class="op">=</span> <span class="va">full</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">leiden_totalVI</span></span>
<span><span class="va">blabs</span> <span class="op">=</span> <span class="va">full</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">batch</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cllabs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## cllabs</span></span>
<span><span class="co">##    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15 </span></span>
<span><span class="co">## 2287 1899 1138  866  787  660  637  587  461  375  334  261  260  208   59   19 </span></span>
<span><span class="co">##   16 </span></span>
<span><span class="co">##   11</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">um</span> <span class="op">=</span> <span class="va">full</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"X_umap"</span><span class="op">)</span> </span>
<span><span class="va">dd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>umap1<span class="op">=</span><span class="va">um</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, umap2<span class="op">=</span><span class="va">um</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, clust<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cllabs</span><span class="op">)</span>, batch<span class="op">=</span><span class="va">blabs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dd</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">umap1</span>, y<span class="op">=</span><span class="va">umap2</span>, colour<span class="op">=</span><span class="va">clust</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citeseq_tut_files/figure-html/getproj-1.png" width="700"></p>
<p>Effectiveness at accommodating the two-batch design is suggested by the mixed representation of the batches in all the Leiden clusters.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dd</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">umap1</span>, y<span class="op">=</span><span class="va">umap2</span>, colour<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="citeseq_tut_files/figure-html/getba-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="protein-abundances-in-projected-clusters">Protein abundances in projected clusters<a class="anchor" aria-label="anchor" href="#protein-abundances-in-projected-clusters"></a>
</h3>
<p>We focus on four of the ADT. Points (cells) in the UMAP projection given above are colored by the estimated abundance of the proteins quantified via (normalized) ADT abundance. Complementary expression of CD4 and CD8a is suggested by the configurations in the middle two panels.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pro4</span> <span class="op">=</span> <span class="va">denoised</span><span class="op">$</span><span class="va">prot_nmlzd</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pro4</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_.*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pro4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">wprot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">dd</span>, <span class="va">pro4</span><span class="op">)</span></span>
<span><span class="va">mm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">wprot</span>, id.vars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clust"</span>, <span class="st">"batch"</span>, <span class="st">"umap1"</span>, <span class="st">"umap2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mm</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   clust   batch     umap1    umap2 variable      value</span></span>
<span><span class="co">## 1     0 PBMC10k 1.0286884 14.58188      CD3  0.6282051</span></span>
<span><span class="co">## 2     0 PBMC10k 0.8443313 15.29502      CD3 11.9159498</span></span>
<span><span class="co">## 3     0 PBMC10k 0.8057814 13.96269      CD3  8.9384489</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mm</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">umap1</span>, y<span class="op">=</span><span class="va">umap2</span>, colour<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">.1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">variable</span><span class="op">)</span></span></code></pre></div>
<p><img src="citeseq_tut_files/figure-html/lknn-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>We have shown that all the results of the totalVI application in the tutorial are readily accessible with utilities in scviR. Additional work on details of differential expression are present in the tutorial and can be explored by the interested reader/user.</p>
</div>
<div class="section level2">
<h2 id="appendix-the-vae-module">Appendix: The VAE module<a class="anchor" aria-label="anchor" href="#appendix-the-vae-module"></a>
</h2>
<p>The structure of the VAE is reported using</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vae</span><span class="op">$</span><span class="va">module</span></span></code></pre></div>
<pre><code><span><span class="co">## TOTALVAE(</span></span>
<span><span class="co">##   (encoder): EncoderTOTALVI(</span></span>
<span><span class="co">##     (encoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=4016, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##         (Layer 1): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=258, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (z_mean_encoder): Linear(in_features=256, out_features=20, bias=True)</span></span>
<span><span class="co">##     (z_var_encoder): Linear(in_features=256, out_features=20, bias=True)</span></span>
<span><span class="co">##     (l_gene_encoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=4016, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (l_gene_mean_encoder): Linear(in_features=256, out_features=1, bias=True)</span></span>
<span><span class="co">##     (l_gene_var_encoder): Linear(in_features=256, out_features=1, bias=True)</span></span>
<span><span class="co">##   )</span></span>
<span><span class="co">##   (decoder): DecoderTOTALVI(</span></span>
<span><span class="co">##     (px_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=22, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (px_scale_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=4000, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): None</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (px_scale_activation): Softmax(dim=-1)</span></span>
<span><span class="co">##     (py_back_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=22, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (py_back_mean_log_alpha): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=14, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): None</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (py_back_mean_log_beta): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=14, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): None</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (py_fore_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=22, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (py_fore_scale_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=14, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (sigmoid_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=22, out_features=256, bias=True)</span></span>
<span><span class="co">##           (1): BatchNorm1d(256, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): ReLU()</span></span>
<span><span class="co">##           (4): Dropout(p=0.2, inplace=False)</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (px_dropout_decoder_gene): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=4000, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): None</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##     (py_background_decoder): FCLayers(</span></span>
<span><span class="co">##       (fc_layers): Sequential(</span></span>
<span><span class="co">##         (Layer 0): Sequential(</span></span>
<span><span class="co">##           (0): Linear(in_features=278, out_features=14, bias=True)</span></span>
<span><span class="co">##           (1): None</span></span>
<span><span class="co">##           (2): None</span></span>
<span><span class="co">##           (3): None</span></span>
<span><span class="co">##           (4): None</span></span>
<span><span class="co">##         )</span></span>
<span><span class="co">##       )</span></span>
<span><span class="co">##     )</span></span>
<span><span class="co">##   )</span></span>
<span><span class="co">## )</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 Patched (2023-08-27 r85021)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /home/vincent/R-4-3-dist/lib/R/lib/libRblas.so </span></span>
<span><span class="co">## LAPACK: /home/vincent/R-4-3-dist/lib/R/lib/libRlapack.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats    graphics utils    stats4   methods  base    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] reshape2_1.4.4              ggplot2_3.4.3              </span></span>
<span><span class="co">##  [3] scviR_1.1.1                 SingleCellExperiment_1.23.0</span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.31.1 Biobase_2.61.0             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.53.1        GenomeInfoDb_1.37.4        </span></span>
<span><span class="co">##  [9] IRanges_2.35.2              S4Vectors_0.39.1           </span></span>
<span><span class="co">## [11] BiocGenerics_0.47.0         MatrixGenerics_1.13.1      </span></span>
<span><span class="co">## [13] matrixStats_1.0.0           shiny_1.7.5                </span></span>
<span><span class="co">## [15] basilisk_1.13.1             BiocStyle_2.29.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.1.3                 bitops_1.0-7             </span></span>
<span><span class="co">##   [3] gridExtra_2.3             rlang_1.1.1              </span></span>
<span><span class="co">##   [5] magrittr_2.0.3            scater_1.29.4            </span></span>
<span><span class="co">##   [7] compiler_4.3.1            RSQLite_2.3.1            </span></span>
<span><span class="co">##   [9] dir.expiry_1.9.0          DelayedMatrixStats_1.23.4</span></span>
<span><span class="co">##  [11] png_0.1-8                 systemfonts_1.0.4        </span></span>
<span><span class="co">##  [13] vctrs_0.6.3               stringr_1.5.0            </span></span>
<span><span class="co">##  [15] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">##  [17] fastmap_1.1.1             dbplyr_2.3.3             </span></span>
<span><span class="co">##  [19] XVector_0.41.1            ellipsis_0.3.2           </span></span>
<span><span class="co">##  [21] labeling_0.4.3            scuttle_1.11.2           </span></span>
<span><span class="co">##  [23] utf8_1.2.3                promises_1.2.1           </span></span>
<span><span class="co">##  [25] rmarkdown_2.24            grDevices_4.3.1          </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.7.2          ragg_1.2.5               </span></span>
<span><span class="co">##  [29] purrr_1.0.2               bit_4.0.5                </span></span>
<span><span class="co">##  [31] xfun_0.40                 zlibbioc_1.47.0          </span></span>
<span><span class="co">##  [33] cachem_1.0.8              beachmat_2.17.16         </span></span>
<span><span class="co">##  [35] jsonlite_1.8.7            blob_1.2.4               </span></span>
<span><span class="co">##  [37] later_1.3.1               DelayedArray_0.27.10     </span></span>
<span><span class="co">##  [39] BiocParallel_1.35.4       irlba_2.3.5.1            </span></span>
<span><span class="co">##  [41] parallel_4.3.1            R6_2.5.1                 </span></span>
<span><span class="co">##  [43] RColorBrewer_1.1-3        bslib_0.5.1              </span></span>
<span><span class="co">##  [45] stringi_1.7.12            limma_3.57.7             </span></span>
<span><span class="co">##  [47] reticulate_1.32.0         jquerylib_0.1.4          </span></span>
<span><span class="co">##  [49] Rcpp_1.0.11               bookdown_0.35            </span></span>
<span><span class="co">##  [51] knitr_1.44                httpuv_1.6.11            </span></span>
<span><span class="co">##  [53] Matrix_1.6-1              tidyselect_1.2.0         </span></span>
<span><span class="co">##  [55] viridis_0.6.4             abind_1.4-5              </span></span>
<span><span class="co">##  [57] yaml_2.3.7                codetools_0.2-19         </span></span>
<span><span class="co">##  [59] curl_5.0.2                plyr_1.8.8               </span></span>
<span><span class="co">##  [61] lattice_0.21-8            tibble_3.2.1             </span></span>
<span><span class="co">##  [63] withr_2.5.0               basilisk.utils_1.13.3    </span></span>
<span><span class="co">##  [65] evaluate_0.21             desc_1.4.2               </span></span>
<span><span class="co">##  [67] BiocFileCache_2.9.1       pillar_1.9.0             </span></span>
<span><span class="co">##  [69] BiocManager_1.30.22       filelock_1.0.2           </span></span>
<span><span class="co">##  [71] generics_0.1.3            rprojroot_2.0.3          </span></span>
<span><span class="co">##  [73] RCurl_1.98-1.12           munsell_0.5.0            </span></span>
<span><span class="co">##  [75] sparseMatrixStats_1.13.4  scales_1.2.1             </span></span>
<span><span class="co">##  [77] xtable_1.8-4              glue_1.6.2               </span></span>
<span><span class="co">##  [79] pheatmap_1.0.12           tools_4.3.1              </span></span>
<span><span class="co">##  [81] datasets_4.3.1            BiocNeighbors_1.19.0     </span></span>
<span><span class="co">##  [83] ScaledMatrix_1.9.1        fs_1.6.3                 </span></span>
<span><span class="co">##  [85] grid_4.3.1                colorspace_2.1-0         </span></span>
<span><span class="co">##  [87] GenomeInfoDbData_1.2.10   beeswarm_0.4.0           </span></span>
<span><span class="co">##  [89] BiocSingular_1.17.1       vipor_0.4.5              </span></span>
<span><span class="co">##  [91] cli_3.6.1                 rsvd_1.0.5               </span></span>
<span><span class="co">##  [93] textshaping_0.3.6         fansi_1.0.4              </span></span>
<span><span class="co">##  [95] viridisLite_0.4.2         S4Arrays_1.1.6           </span></span>
<span><span class="co">##  [97] dplyr_1.1.3               gtable_0.3.4             </span></span>
<span><span class="co">##  [99] sass_0.4.7                digest_0.6.33            </span></span>
<span><span class="co">## [101] ggrepel_0.9.3             SparseArray_1.1.12       </span></span>
<span><span class="co">## [103] farver_2.1.1              memoise_2.0.1            </span></span>
<span><span class="co">## [105] htmltools_0.5.6           pkgdown_2.0.7            </span></span>
<span><span class="co">## [107] lifecycle_1.0.3           httr_1.4.7               </span></span>
<span><span class="co">## [109] statmod_1.5.0             mime_0.12                </span></span>
<span><span class="co">## [111] bit64_4.0.5</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
