<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing totalVI and OSCA book CITE-seq analyses • scviR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing totalVI and OSCA book CITE-seq analyses">
<meta property="og:description" content="scviR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scviR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scviR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/citeseq_tut.html">scvi-tools CITE-seq tutorial in R, using serialized tutorial components</a>
    </li>
    <li>
      <a href="../articles/compch12.html">Comparing totalVI and OSCA book CITE-seq analyses</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/scviR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing totalVI and OSCA book CITE-seq
analyses</h1>
                        <h4 data-toc-skip class="author">OSCA book
authors and Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">March 26, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/scviR/blob/HEAD/vignettes/compch12.Rmd"><code>vignettes/compch12.Rmd</code></a></small>
      <div class="hidden name"><code>compch12.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette endeavors to put Bioconductor and scvi-tools together
to help understand how different data structures and methods relevant to
CITE-seq analysis contribute to interpretation of CITE-seq
exeperiments.</p>
<p>The scvi-tools tutorial (for version 0.20.0) analyzes a pair of 10x
PBMC CITE-seq experiments (5k and 10k cells). Chapter 12 of the OSCA
book analyzes only the 10k dataset.</p>
</div>
<div class="section level2">
<h2 id="technical-steps-to-facilitate-comparison">Technical steps to facilitate comparison<a class="anchor" aria-label="anchor" href="#technical-steps-to-facilitate-comparison"></a>
</h2>
<p>The following subsections are essentially “code-only”. We exhibit
steps necessary to assemble a SingleCellExperiment instance with the
subset of the totalVI quantifications produced for the cells from the
“10k” dataset.</p>
<div class="section level3">
<h3 id="acquire-software">Acquire software<a class="anchor" aria-label="anchor" href="#acquire-software"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/scviR">scviR</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="obtain-key-data-representations">Obtain key data representations<a class="anchor" aria-label="anchor" href="#obtain-key-data-representations"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ch12sce</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCh12Sce.html">getCh12Sce</a></span><span class="op">(</span>clear_cache<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ch12sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 33538 7472 </span></span>
<span><span class="co">## metadata(2): Samples se.averaged</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span>
<span><span class="co">##   ENSG00000268674</span></span>
<span><span class="co">## rowData names(3): ID Symbol Type</span></span>
<span><span class="co">## colnames(7472): AAACCCAAGATTGTGA-1 AAACCCACATCGGTTA-1 ...</span></span>
<span><span class="co">##   TTTGTTGTCGAGTGAG-1 TTTGTTGTCGTTCAGA-1</span></span>
<span><span class="co">## colData names(3): Sample Barcode sizeFactor</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: Gene Expression</span></span>
<span><span class="co">## altExpNames(1): Antibody Capture</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fullvi</span> <span class="op">=</span> <span class="fu"><a href="../reference/getTotalVI5k10kAdata.html">getTotalVI5k10kAdata</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fullvi</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 10849 × 4000</span></span>
<span><span class="co">##     obs: 'n_genes', 'percent_mito', 'n_counts', 'batch', '_scvi_labels', '_scvi_batch', 'leiden_totalVI'</span></span>
<span><span class="co">##     var: 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'</span></span>
<span><span class="co">##     uns: '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'leiden', 'log1p', 'neighbors', 'umap'</span></span>
<span><span class="co">##     obsm: 'X_totalVI', 'X_umap', 'denoised_protein', 'protein_expression', 'protein_foreground_prob'</span></span>
<span><span class="co">##     layers: 'counts', 'denoised_rna'</span></span>
<span><span class="co">##     obsp: 'connectivities', 'distances'</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="assemble-a-singlecellexperiment-with-totalvi-outputs">Assemble a SingleCellExperiment with totalVI outputs<a class="anchor" aria-label="anchor" href="#assemble-a-singlecellexperiment-with-totalvi-outputs"></a>
</h3>
<div class="section level4">
<h4 id="acquire-cell-identities-and-batch-labels">Acquire cell identities and batch labels<a class="anchor" aria-label="anchor" href="#acquire-cell-identities-and-batch-labels"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">totvi_cellids</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fullvi</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span></span>
<span><span class="va">totvi_batch</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">batch</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="acquire-quantifications-and-latent-space-positions">Acquire quantifications and latent space positions<a class="anchor" aria-label="anchor" href="#acquire-quantifications-and-latent-space-positions"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">totvi_latent</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"X_totalVI"</span><span class="op">)</span></span>
<span><span class="va">totvi_umap</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"X_umap"</span><span class="op">)</span></span>
<span><span class="va">totvi_denoised_rna</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">layers</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"denoised_rna"</span><span class="op">)</span></span>
<span><span class="va">totvi_denoised_protein</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"denoised_protein"</span><span class="op">)</span></span>
<span><span class="va">totvi_leiden</span> <span class="op">=</span> <span class="va">fullvi</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">leiden_totalVI</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="drop-5k-data-from-all">Drop 5k data from all<a class="anchor" aria-label="anchor" href="#drop-5k-data-from-all"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is5k</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">totvi_batch</span> <span class="op">==</span> <span class="st">"PBMC5k"</span><span class="op">)</span></span>
<span><span class="va">totvi_cellids</span> <span class="op">=</span> <span class="va">totvi_cellids</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span><span class="op">]</span></span>
<span><span class="va">totvi_latent</span> <span class="op">=</span> <span class="va">totvi_latent</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span>,<span class="op">]</span></span>
<span><span class="va">totvi_umap</span> <span class="op">=</span> <span class="va">totvi_umap</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span>,<span class="op">]</span></span>
<span><span class="va">totvi_denoised_rna</span> <span class="op">=</span> <span class="va">totvi_denoised_rna</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span>,<span class="op">]</span></span>
<span><span class="va">totvi_denoised_protein</span> <span class="op">=</span> <span class="va">totvi_denoised_protein</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span>,<span class="op">]</span></span>
<span><span class="va">totvi_leiden</span> <span class="op">=</span> <span class="va">totvi_leiden</span><span class="op">[</span><span class="op">-</span><span class="va">is5k</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="label-the-rows-of-components">Label the rows of components<a class="anchor" aria-label="anchor" href="#label-the-rows-of-components"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totvi_latent</span><span class="op">)</span> <span class="op">=</span> <span class="va">totvi_cellids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totvi_umap</span><span class="op">)</span> <span class="op">=</span> <span class="va">totvi_cellids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totvi_denoised_rna</span><span class="op">)</span> <span class="op">=</span> <span class="va">totvi_cellids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totvi_denoised_protein</span><span class="op">)</span> <span class="op">=</span> <span class="va">totvi_cellids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">totvi_leiden</span><span class="op">)</span> <span class="op">=</span> <span class="va">totvi_cellids</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="find-common-cell-ids">Find common cell ids<a class="anchor" aria-label="anchor" href="#find-common-cell-ids"></a>
</h4>
<p>In this section we reduce the cell collections to cells common to the
chapter 12 and totalVI datasets.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">totvi_cellids</span>, <span class="va">ch12sce</span><span class="op">$</span><span class="va">Barcode</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="build-the-totalvi-singlecellexperiment">Build the totalVI SingleCellExperiment<a class="anchor" aria-label="anchor" href="#build-the-totalvi-singlecellexperiment"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select and order</span></span>
<span><span class="va">totvi_latent</span> <span class="op">=</span> <span class="va">totvi_latent</span><span class="op">[</span><span class="va">comm</span>,<span class="op">]</span></span>
<span><span class="va">totvi_umap</span> <span class="op">=</span> <span class="va">totvi_umap</span><span class="op">[</span><span class="va">comm</span>,<span class="op">]</span></span>
<span><span class="va">totvi_denoised_rna</span> <span class="op">=</span> <span class="va">totvi_denoised_rna</span><span class="op">[</span><span class="va">comm</span>,<span class="op">]</span></span>
<span><span class="va">totvi_denoised_protein</span> <span class="op">=</span> <span class="va">totvi_denoised_protein</span><span class="op">[</span><span class="va">comm</span>,<span class="op">]</span></span>
<span><span class="va">totvi_leiden</span> <span class="op">=</span> <span class="va">totvi_leiden</span><span class="op">[</span><span class="va">comm</span><span class="op">]</span></span>
<span> </span>
<span><span class="co"># organize the totalVI into SCE with altExp</span></span>
<span></span>
<span><span class="va">totsce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu">SimpleList</span><span class="op">(</span>logcounts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">totvi_denoised_rna</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># FALSE name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span> <span class="op">=</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="va">fullvi</span><span class="op">$</span><span class="va">var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fullvi</span><span class="op">$</span><span class="va">var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span></span>
<span><span class="va">nn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu">SimpleList</span><span class="op">(</span>logcounts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">totvi_denoised_protein</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># FALSE name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">nn</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>UMAP<span class="op">=</span><span class="va">totvi_umap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span> <span class="op">=</span> <span class="va">nn</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExpNames</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span> <span class="op">=</span> <span class="st">"denoised_protein"</span></span>
<span><span class="va">totsce</span><span class="op">$</span><span class="va">leiden</span> <span class="op">=</span> <span class="va">totvi_leiden</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span> <span class="op">=</span> <span class="va">totvi_leiden</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">ch12.clusters</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">ch12sce</span><span class="op">[</span>,<span class="va">comm</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">label</span></span>
<span></span>
<span><span class="co"># add average ADT abundance to metadata, for adt_profiles</span></span>
<span></span>
<span><span class="va">tot.se.averaged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/sumCountsAcrossCells.html" class="external-link">sumCountsAcrossCells</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span>,</span>
<span>    exprs_values<span class="op">=</span><span class="st">"logcounts"</span>, average<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tot.se.averaged</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_TotalSeqB"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tot.se.averaged</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">se.averaged</span> <span class="op">=</span> <span class="va">tot.se.averaged</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reduce-the-chapter-12-dataset-to-the-cells-held-in-common">Reduce the chapter 12 dataset to the cells held in common<a class="anchor" aria-label="anchor" href="#reduce-the-chapter-12-dataset-to-the-cells-held-in-common"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ch12sce</span><span class="op">)</span> <span class="op">=</span> <span class="va">ch12sce</span><span class="op">$</span><span class="va">Barcode</span></span>
<span><span class="va">ch12sce_matched</span> <span class="op">=</span> <span class="va">ch12sce</span><span class="op">[</span>, <span class="va">comm</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="key-outputs-of-the-chapter-12-analysis">Key outputs of the chapter 12 analysis<a class="anchor" aria-label="anchor" href="#key-outputs-of-the-chapter-12-analysis"></a>
</h2>
<div class="section level3">
<h3 id="clustering-and-projection-based-on-the-adt-quantifications">Clustering and projection based on the ADT quantifications<a class="anchor" aria-label="anchor" href="#clustering-and-projection-based-on-the-adt-quantifications"></a>
</h3>
<p>The TSNE projection of the normalized ADT quantifications and the <a href="https://arxiv.org/abs/physics/0512106" class="external-link">walktrap</a> cluster
assignments are produced for the cells common to the two approaches.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">ch12sce_matched</span><span class="op">)</span>, color_by<span class="op">=</span><span class="st">"label"</span>, text_by<span class="op">=</span><span class="st">"label"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 1 rows containing missing values (`geom_text_repel()`).</span></span></code></pre>
<p><img src="compch12_files/figure-html/lkkey-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="cluster-profiles-based-on-averaging-adt-quantities-within-clusters">Cluster profiles based on averaging ADT quantities within
clusters<a class="anchor" aria-label="anchor" href="#cluster-profiles-based-on-averaging-adt-quantities-within-clusters"></a>
</h3>
<p>This heatmap uses precomputed cluster averages that are lodged in the
metadata element of the SingleCellExperiment. Colors represent the
log2-fold change from the grand average across all clusters.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/adtProfiles.html">adtProfiles</a></span><span class="op">(</span><span class="va">ch12sce_matched</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/lkadt-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="marker-expression-patterns-in-mrna-based-sub-clusters-of-adt-based-clusters">Marker expression patterns in mRNA-based sub-clusters of ADT-based
clusters<a class="anchor" aria-label="anchor" href="#marker-expression-patterns-in-mrna-based-sub-clusters-of-adt-based-clusters"></a>
</h3>
<p>We enhance the annotation of the list of subclusters retrieved using
<code>getCh12AllSce</code> and then drill into mRNA-based subclusters of
ADT-based cluster 3 to compare expression levels of three genes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ch12_allsce</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCh12AllSce.html">getCh12AllSce</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">ch12_allsce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ch12_allsce</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Barcode</span>; </span>
<span>   <span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>; </span>
<span>   <span class="va">x</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">cn</span>,<span class="va">comm</span><span class="op">)</span><span class="op">]</span>; <span class="va">x</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">of.interest</span> <span class="op">&lt;-</span> <span class="st">"3"</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GZMH"</span>, <span class="st">"IL7R"</span>, <span class="st">"KLRB1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">ch12_allsce</span><span class="op">[[</span><span class="va">of.interest</span><span class="op">]</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"subcluster"</span>,</span>
<span>    features<span class="op">=</span><span class="va">markers</span>, swap_rownames<span class="op">=</span><span class="st">"Symbol"</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/dosubcl-1.png" width="700"></p>
<p>There is a suggestion of a boolean basis for subcluster identity,
depending on low or high expression of the selected genes.</p>
</div>
<div class="section level3">
<h3 id="graduated-relationships-between-mrna-and-surface-protein-expression">Graduated relationships between mRNA and surface protein
expression<a class="anchor" aria-label="anchor" href="#graduated-relationships-between-mrna-and-surface-protein-expression"></a>
</h3>
<p>Following the exploration in OSCA chapter 12, cluster 3 is analyzed
for a regression association between expression measures of three genes
and the ADT-based abundance of CD127.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">ch12_allsce</span><span class="op">[[</span><span class="st">"3"</span><span class="op">]</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"CD127"</span>, show_smooth<span class="op">=</span><span class="cn">TRUE</span>, show_se<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>    features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IL7R"</span>, <span class="st">"TPT1"</span>, <span class="st">"KLRB1"</span>, <span class="st">"GZMH"</span><span class="op">)</span>, swap_rownames<span class="op">=</span><span class="st">"Symbol"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `geom_smooth()` using method = 'loess' and formula = 'y ~ x'</span></span></code></pre>
<p><img src="compch12_files/figure-html/lkgra-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="analogs-to-the-chapter-12-findings-based-on-totalvi-quantifications">Analogs to the chapter 12 findings, based on totalVI
quantifications<a class="anchor" aria-label="anchor" href="#analogs-to-the-chapter-12-findings-based-on-totalvi-quantifications"></a>
</h2>
<div class="section level3">
<h3 id="the-leiden-clustering-in-umap-projection">The Leiden clustering in UMAP projection<a class="anchor" aria-label="anchor" href="#the-leiden-clustering-in-umap-projection"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span>, color_by<span class="op">=</span><span class="st">"leiden"</span>, text_by<span class="op">=</span><span class="st">"leiden"</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/lktotumap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="cluster-profiles-based-on-average-adt-abundance-using-denoised-protein-quantifications">Cluster profiles based on average ADT abundance, using denoised
protein quantifications<a class="anchor" aria-label="anchor" href="#cluster-profiles-based-on-average-adt-abundance-using-denoised-protein-quantifications"></a>
</h3>
<p>The approach to profiling the ADT abundances used in the totalVI
tutorial employs scaling to (0,1).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tav</span> <span class="op">=</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">se.averaged</span></span>
<span><span class="va">ata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">tav</span><span class="op">)</span></span>
<span><span class="va">uscale</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">scmat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">ata</span>,<span class="fl">1</span>,<span class="va">uscale</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">scmat</span>, cluster_rows<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/domattot-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="concordance-in-adt-based-clustering-between-osca-and-totalvi">Concordance in ADT-based clustering between OSCA and totalVI<a class="anchor" aria-label="anchor" href="#concordance-in-adt-based-clustering-between-osca-and-totalvi"></a>
</h3>
<p>A quick view of the concordance of the two clustering outcomes is</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span></span>
<span><span class="va">ach12</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">ch12sce_matched</span><span class="op">)</span></span>
<span><span class="va">tt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>ch12<span class="op">=</span><span class="va">ach12</span><span class="op">$</span><span class="va">label</span>, VI<span class="op">=</span><span class="va">atot</span><span class="op">$</span><span class="va">leiden</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">tt</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/lkconc-1.png" width="700"></p>
<p>With this we can pick out some clusters with many cells in
common:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lit</span> <span class="op">=</span> <span class="va">tt</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"9"</span>, <span class="st">"12"</span>, <span class="st">"5"</span>, <span class="st">"3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"8"</span>, <span class="st">"6"</span>, <span class="st">"5"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lit</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sQuote.html" class="external-link">sQuote</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lit</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lit</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sQuote.html" class="external-link">sQuote</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lit</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lit</span></span></code></pre></div>
<pre><code><span><span class="co">##       VI</span></span>
<span><span class="co">## ch12    '0'  '1'  '2'  '8'  '6'  '5'</span></span>
<span><span class="co">##   '9'  1334    0    0    0    0    0</span></span>
<span><span class="co">##   '12'    0  993    0    0   15    0</span></span>
<span><span class="co">##   '5'     0  102  671    8   44    2</span></span>
<span><span class="co">##   '3'     0    0    5  322  297   67</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="subcluster-assessment-for-osca-cluster-3">Subcluster assessment for OSCA cluster “3”<a class="anchor" aria-label="anchor" href="#subcluster-assessment-for-osca-cluster-3"></a>
</h3>
<p>Let’s examine the distributions of marker mRNAs in the Leiden totalVI
clusters corresponding to OSCA cluster “3”:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tsub</span> <span class="op">=</span> <span class="va">totsce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">totsce</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"8"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GZMH"</span>, <span class="st">"IL7R"</span>, <span class="st">"KLRB1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span><span class="op">)</span> <span class="co"># squelch unused levels</span></span>
<span><span class="va">tsub</span><span class="op">$</span><span class="va">leiden</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">$</span><span class="va">leiden</span><span class="op">)</span> <span class="co"># squelch unused levels</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">tsub</span>, x<span class="op">=</span><span class="st">"leiden"</span>,</span>
<span>    features<span class="op">=</span><span class="va">markers</span>, swap_rownames<span class="op">=</span><span class="st">"Symbol"</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="compch12_files/figure-html/lksubsss-1.png" width="700"></p>
<p>Note that the y axis label is incorrect – we are plotting the
denoised expression values from totalVI.</p>
<p>The display seems roughly consistent with the “boolean basis”
observed above with the mRNA-based subclustering.</p>
</div>
<div class="section level3">
<h3 id="graduated-relationships-between-adt-and-mrna-abundance-as-measured-by-totalvi">Graduated relationships between ADT and mRNA abundance as measured
by totalVI<a class="anchor" aria-label="anchor" href="#graduated-relationships-between-adt-and-mrna-abundance-as-measured-by-totalvi"></a>
</h3>
<p>The same approach is taken as above. We don’t have TPT1 in the 4000
genes retained in the totalVI exercise.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_TotalSeqB"</span>, <span class="st">""</span>, <span class="va">rn</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">)</span><span class="op">)</span> <span class="op">=</span> <span class="va">rn</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">tsub</span><span class="op">)</span><span class="op">)</span> <span class="op">=</span> <span class="fu">DataFrame</span><span class="op">(</span>Symbol<span class="op">=</span><span class="va">rn</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">tsub</span>, x<span class="op">=</span><span class="st">"CD127"</span>, show_smooth<span class="op">=</span><span class="cn">TRUE</span>, show_se<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>   features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IL7R"</span>, <span class="st">"KLRB1"</span>, <span class="st">"GZMH"</span><span class="op">)</span>, swap_rownames<span class="op">=</span><span class="st">"Symbol"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'</span></span></code></pre>
<p><img src="compch12_files/figure-html/lkgradvi-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>We have shown how rudimentary programming and data organization can
be used to make outputs of OSCA and totalVI methods amenable to
comparison in the Bioconductor framework.</p>
<p>The scviR package includes a shiny app in the function
<code>explore_subcl</code> that should be expanded to facilitate
exploration of totalVI subclusters. Much more work remains to be done in
the area of exploring</p>
<ul>
<li><p>additional approaches to integrative interpretation of ADT and
mRNA abundance patterns, such as intersection and concatenation methods
in the feature selection materials in OSCA ch. 12</p></li>
<li><p>effects of tuning and architecture details for the totalVI
VAE</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2023-03-16 r83996)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats    graphics utils    stats4   methods  base    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scviR_0.99.10               shiny_1.7.4                </span></span>
<span><span class="co">##  [3] basilisk_1.11.3             scater_1.27.8              </span></span>
<span><span class="co">##  [5] ggplot2_3.4.1               scuttle_1.9.4              </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.21.1 SummarizedExperiment_1.29.1</span></span>
<span><span class="co">##  [9] Biobase_2.59.0              GenomicRanges_1.51.4       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.35.16        IRanges_2.33.0             </span></span>
<span><span class="co">## [13] S4Vectors_0.37.4            BiocGenerics_0.45.2        </span></span>
<span><span class="co">## [15] MatrixGenerics_1.11.0       matrixStats_0.63.0         </span></span>
<span><span class="co">## [17] BiocStyle_2.27.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.1.3                 bitops_1.0-7             </span></span>
<span><span class="co">##   [3] gridExtra_2.3             rlang_1.1.0              </span></span>
<span><span class="co">##   [5] magrittr_2.0.3            RSQLite_2.3.0            </span></span>
<span><span class="co">##   [7] compiler_4.3.0            mgcv_1.8-42              </span></span>
<span><span class="co">##   [9] dir.expiry_1.7.0          DelayedMatrixStats_1.21.0</span></span>
<span><span class="co">##  [11] png_0.1-8                 systemfonts_1.0.4        </span></span>
<span><span class="co">##  [13] vctrs_0.6.1               stringr_1.5.0            </span></span>
<span><span class="co">##  [15] pkgconfig_2.0.3           fastmap_1.1.1            </span></span>
<span><span class="co">##  [17] dbplyr_2.3.2              XVector_0.39.0           </span></span>
<span><span class="co">##  [19] ellipsis_0.3.2            labeling_0.4.2           </span></span>
<span><span class="co">##  [21] utf8_1.2.3                promises_1.2.0.1         </span></span>
<span><span class="co">##  [23] rmarkdown_2.20            grDevices_4.3.0          </span></span>
<span><span class="co">##  [25] ggbeeswarm_0.7.1          ragg_1.2.5               </span></span>
<span><span class="co">##  [27] bit_4.0.5                 purrr_1.0.1              </span></span>
<span><span class="co">##  [29] xfun_0.38                 zlibbioc_1.45.0          </span></span>
<span><span class="co">##  [31] cachem_1.0.7              beachmat_2.15.0          </span></span>
<span><span class="co">##  [33] jsonlite_1.8.4            blob_1.2.4               </span></span>
<span><span class="co">##  [35] highr_0.10                later_1.3.0              </span></span>
<span><span class="co">##  [37] DelayedArray_0.25.0       BiocParallel_1.33.11     </span></span>
<span><span class="co">##  [39] irlba_2.3.5.1             parallel_4.3.0           </span></span>
<span><span class="co">##  [41] R6_2.5.1                  RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [43] bslib_0.4.2               stringi_1.7.12           </span></span>
<span><span class="co">##  [45] limma_3.55.5              reticulate_1.28          </span></span>
<span><span class="co">##  [47] jquerylib_0.1.4           Rcpp_1.0.10              </span></span>
<span><span class="co">##  [49] bookdown_0.33             knitr_1.42               </span></span>
<span><span class="co">##  [51] splines_4.3.0             httpuv_1.6.9             </span></span>
<span><span class="co">##  [53] Matrix_1.5-3              tidyselect_1.2.0         </span></span>
<span><span class="co">##  [55] yaml_2.3.7                viridis_0.6.2            </span></span>
<span><span class="co">##  [57] codetools_0.2-19          curl_5.0.0               </span></span>
<span><span class="co">##  [59] lattice_0.20-45           tibble_3.2.1             </span></span>
<span><span class="co">##  [61] basilisk.utils_1.11.2     withr_2.5.0              </span></span>
<span><span class="co">##  [63] evaluate_0.20             desc_1.4.2               </span></span>
<span><span class="co">##  [65] BiocFileCache_2.7.2       pillar_1.9.0             </span></span>
<span><span class="co">##  [67] BiocManager_1.30.20       filelock_1.0.2           </span></span>
<span><span class="co">##  [69] generics_0.1.3            rprojroot_2.0.3          </span></span>
<span><span class="co">##  [71] RCurl_1.98-1.10           sparseMatrixStats_1.11.1 </span></span>
<span><span class="co">##  [73] munsell_0.5.0             scales_1.2.1             </span></span>
<span><span class="co">##  [75] xtable_1.8-4              glue_1.6.2               </span></span>
<span><span class="co">##  [77] pheatmap_1.0.12           tools_4.3.0              </span></span>
<span><span class="co">##  [79] datasets_4.3.0            BiocNeighbors_1.17.1     </span></span>
<span><span class="co">##  [81] ScaledMatrix_1.7.0        fs_1.6.1                 </span></span>
<span><span class="co">##  [83] grid_4.3.0                colorspace_2.1-0         </span></span>
<span><span class="co">##  [85] nlme_3.1-162              GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">##  [87] beeswarm_0.4.0            BiocSingular_1.15.0      </span></span>
<span><span class="co">##  [89] vipor_0.4.5               cli_3.6.1                </span></span>
<span><span class="co">##  [91] rsvd_1.0.5                textshaping_0.3.6        </span></span>
<span><span class="co">##  [93] fansi_1.0.4               viridisLite_0.4.1        </span></span>
<span><span class="co">##  [95] dplyr_1.1.1               gtable_0.3.3             </span></span>
<span><span class="co">##  [97] sass_0.4.5                digest_0.6.31            </span></span>
<span><span class="co">##  [99] ggrepel_0.9.3             farver_2.1.1             </span></span>
<span><span class="co">## [101] memoise_2.0.1             htmltools_0.5.5          </span></span>
<span><span class="co">## [103] pkgdown_2.0.7             lifecycle_1.0.3          </span></span>
<span><span class="co">## [105] here_1.0.1                httr_1.4.5               </span></span>
<span><span class="co">## [107] mime_0.12                 bit64_4.0.5</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
